### PR TITLE
Add multi_tensor_adam_param_remainder and context parallel support

### PR DESCRIPTION
**Pull request by lihongyang1990**
## Summary
- flagos: Add multi_tensor_adam_param_remainder implementation
- reference: Add multi_tensor_adam_param_remainder implementation  
- reference: Add context parallel support for Flash Attention
- manager: Add cache mechanism with _impl_cache and _impl_cache_meta for conditional op selection

### flagos backend
- Implemented multi_tensor_adam_param_remainder operation for handling parameter remainders in multi-tensor Adam optimizer

### reference backend  
- Implemented multi_tensor_adam_param_remainder operation
- Added context parallel support for Flash Attention implementation

### Core manager
- Added cache mechanism using _impl_cache and _impl_cache_meta
- Improved op selection with conditional caching based on policy fingerprint and epoch

### CODE REVIEW

File: transformer_engine/plugin/core/manager.py [hunk 1]
```diff
@@ -42,7 +42,8 @@ def __init__(self, registry: Optional[OpRegistry] = None) -> None:
         self._registry = registry or OpRegistry()
         self._state = _OpManagerState()
         self._dispatch_cache: Dict[Tuple[str, str, int], Callable] = {}
-        self._called_ops: Dict[str, str] = {}  # Map op_name -> last_used_impl_id (for logging)
+        self._impl_cache: Dict[str, OpImpl] = {}
+        self._impl_cache_meta: Dict[str, Tuple[str, int]] = {}
```
Comment (lxd-cumt) [inline transformer_engine/plugin/core/manager.py:46]: What the role of impl_cache_meta is, and why we need to maintain an impl_cache and impl_cache_meta

File: transformer_engine/plugin/core/backends/reference/flash_attention.py [hunk 2]
```diff
@@ -7,10 +7,78 @@
 
 import torch
 import torch.nn.functional as F
+import torch.distributed as dist
 
 from transformer_engine.plugin.core.ops import FlashAttentionBase
 
 
+def _all_gather_along_seq(
```
Comment (lxd-cumt) [inline transformer_engine/plugin/core/backends/reference/flash_attention.py]: Currently, to support CP, we can implement attention by using CP-group all-gather and reduce-scatter. However, this still performs the full QKV attention computation on every CP rank, which increases memory usage and hurts performance.

In the future we can let each CP rank compute separately and then merge the softmax lse to obtain reference Flash-Attention.

File: transformer_engine/plugin/core/backends/reference/flash_attention.py [hunk 3]
```diff
@@ -7,10 +7,78 @@
...
+            # Split the gradient and take only the local chunk
+            grad_chunks = torch.chunk(grad_output, ctx.world_size, dim=seq_dim)
+
+            # Reduce scatter: sum gradients from all ranks and scatter
+            local_grad = torch.zeros_like(grad_chunks[0])
+            grad_list = [torch.empty_like(grad_chunks[0]) for _ in range(ctx.world_size)]
+            for i, chunk in enumerate(grad_chunks):
+                grad_list[i].copy_(chunk)
+
+            dist.reduce_scatter(local_grad, grad_list, group=ctx.cp_group)
```
Comment (lxd-cumt) [inline transformer_engine/plugin/core/backends/reference/flash_attention.py]: reduce_scatter or scatter?

File: transformer_engine/plugin/core/backends/reference/flash_attention.py [hunk 4]
```diff
@@ -7,10 +7,78 @@
...
+            return local_grad
+
+    return AllGatherFunc.apply(tensor)
+
+
+def _reduce_scatter_along_seq(
```
Comment (lxd-cumt) [inline transformer_engine/plugin/core/backends/reference/flash_attention.py]: Is this used here?

**Review by lxd-cumt — APPROVED** [review]
LGTM

**Review by zhaoyinglia — APPROVED** [review]
LGTM
